src/Reopt/TypeInference/Solver/Monad.hs

```diff
@@ -291,27 +302,29 @@ data Conditional a = Conditional
 -- one exists.
 condEnabled :: Conditional c -> SolverM (Maybe [EqRowC])
 condEnabled c = asum <$> mapM matchAll (cGuard c)
-  where
-    matchAll pats = fmap concat . sequenceA <$> mapM match pats
-    match pat = do
-      (_, m_ty) <- lookupTyVar (patVar pat)
-      pure (matchRHS (patRHS pat) =<< m_ty)
-
-    matchRHS rhs ty =
-      case (rhs, ty) of
-        (IsPtr sv  , PtrTy r)      -> Just (bindSchem sv r)
-        (IsNum     , NumTy _)      -> Just []
-        (IsConflict, ConflictTy _) -> Just []        
-        _                          -> Nothing
-
-    bindSchem (Schematic r) re = [EqRowC (RowExprVar r) re]
-    bindSchem _             _  = []
+ where
+  matchAll (Conjunction pats) = fmap concat . sequenceA <$> mapM match pats
+  match pat = do
+    (_, m_ty) <- lookupTyVar (patVar pat)
+    pure (matchRHS (patRHS pat) =<< m_ty)
+
+  matchRHS rhs ty =
+    case (rhs, ty) of
+      (IsPtr sv, PtrTy r) -> Just (bindSchem sv r)
+      (IsNum, NumTy _) -> Just []
+      (IsConflict, ConflictTy _) -> Just []
+      _ -> Nothing
+
+  bindSchem (Schematic r) re = [EqRowC (RowExprVar r) re]
+  bindSchem _ _ = []
 
 instance PP.Pretty a => PP.Pretty (Conditional a) where
-  pretty c = PP.pretty (cName c) <> ": " <>
-             -- FIXME, could be nicer.
-             PP.list (map (PP.list . map PP.pretty) (cGuard c)) <> " |- " <>
-             PP.pretty (cConstraints c)
+  pretty c =
+    PP.pretty (cName c)
+    <> ": "
+    <> PP.list (getDisjuncts (fmap (PP.list . getConjuncts . fmap PP.pretty) (cGuard c)))
```

Comment:
...this code, which is quite involved.

Reply:
Yeah I'd be happy with enforcing such a `Pretty` instance for these, given their use is very limited, I doubt we'd ever not want `PP.list`.